Dump container logs on shell command failure

pugnascotia opened the pull request:
The Docker tests framework captures container logs when Elasticsearch
fails to start. However, it doesn't do this if a later shell command
fails. Amend the `DockerShell` wrapper to capture the container logs if
a command fails when it should succeed.

pugnascotia committed:
Dump container logs on shell command failure

The Docker tests framework captures container logs when Elasticsearch
fails to start. However, it doesn't do this if a later shell command
fails. Amend the DockerShell wrapper to capture the container logs if
a command fails when it should succeed.

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Packaging)

pugnascotia commented:
Ref: #51316

mark-vieira commented on qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java:
```diff
@@ -291,6 +291,22 @@ public static void copyFromContainer(Path from, Path to) {
+                        e.getMessage() + "\n\nContainer stdout: " + dockerLogs.stdout + "\n\nContainer stderr: " + dockerLogs.stderr
+                    );
+                } catch (ShellException logsException) {
+                    throw new ShellException("Command failed, and couldn't get docker logs", logsException);
```
We lose the original exception in this scenario and effectively replace it with one about not being able to fetch logs. We should probably still rethrow the original exception and maybe add this one as a suppressed exception.

mark-vieira commented on qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java:
```diff
@@ -291,6 +291,22 @@ public static void copyFromContainer(Path from, Path to) {
+                try {
+                    final Shell.Result dockerLogs = getContainerLogs();
+                    throw new ShellException(
+                        e.getMessage() + "\n\nContainer stdout: " + dockerLogs.stdout + "\n\nContainer stderr: " + dockerLogs.stderr
```
Even though we are logging the original exception message, we should still add it as the cause to this one, otherwise we lose the original call site stacktrace.

mark-vieira reviewed (commented):
Couple of comments.

rjernst commented on qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java:
```diff
@@ -291,6 +291,22 @@ public static void copyFromContainer(Path from, Path to) {
+                    throw new ShellException(
+                        e.getMessage() + "\n\nContainer stdout: " + dockerLogs.stdout + "\n\nContainer stderr: " + dockerLogs.stderr
+                    );
+                } catch (ShellException logsException) {
```
Why are we catching an exception we just threw on the line above? It seems like we are adding an additional stack frame, why?

pugnascotia committed:
Throw original exception docker logs capture

When DockerShell.run() catches a ShellException, dump out the logs and
then rethrow the exception, rather than wrapping the original exception.

pugnascotia commented:
I've reworked the implementation here to dump out the container logs but use the same exception, please take another look.

rjernst commented on qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java:
```diff
@@ -291,6 +291,36 @@ public static void copyFromContainer(Path from, Path to) {
+                    logger.error(
+                        "Command [{}] failed.\n\nTried to dump container logs but that failed too: [{}]",
+                        script,
+                        shellException.getMessage()
```
This is losing the stack trace of the nested exception. Instead we should attach this as a suppressed exception to the original exception

mark-vieira reviewed (approved):
Changes LGTM 👍

mark-vieira commented:
@pugnascotia Do you mind backporting this to `7.x` and `7.6`? Seems like it's be a really helpful thing to diagnose failures like [this one](https://gradle-enterprise.elastic.co/s/m44pengwg72qk/tests/zyp3a6atiroz4-eos5vmgj6xf5i).
